src/main/kotlin/dev/sparky200/skriptnbt/skript/expressions/NbtOfHolderExpression.kt

```diff
@@ -0,0 +1,153 @@
+                ExpressionType.PROPERTY,
+                "nbt [compound] (of|from) %objects%",
+                "nbt [compound] (of|from) file[s] %strings%",
+                "nbt [compound] (of|from) text %strings%"
+            )
+        }
+    }
+
+    private var isFromFile: Boolean = false
+    private var isFromText: Boolean = false
+    private val canBeMutated get() = !isFromFile && !isFromText
+
+    private fun rhsError(): Nothing =
+        throw IllegalArgumentException("The right-hand side of the effect must be an nbt compound")
+
+    private fun noNbtError(event: Event?): Nothing =
+        throw IllegalArgumentException("'${expr.toString(event, false)}' does not have NBT")
+
+    override fun acceptChange(mode: Changer.ChangeMode?): Array<Class<*>?>? {
+        if (!canBeMutated) return null
+        return when (mode) {
+            Changer.ChangeMode.ADD,
+            Changer.ChangeMode.SET -> arrayOf(MutableCompound::class.java)
+            Changer.ChangeMode.REMOVE -> arrayOf(MutableCompound::class.java, Collection::class.java, String::class.java, Array::class.java)
+            else -> {
+                super.acceptChange(mode)
+            }
+        }
+    }
+
+    override fun change(event: Event?, delta: Array<out Any?>?, mode: Changer.ChangeMode?) {
+        if (!canBeMutated) throw IllegalStateException("Cannot mutate ")
+        val theDelta =
+            if (delta != null && delta.isNotEmpty())
+                delta[0]
+            else null
+
+        when (mode) {
+            Changer.ChangeMode.SET -> {
+                if (theDelta == null) rhsError()
+                @Suppress("UNCHECKED_CAST")
+                if (expr.isSingle) set(event, expr.getSingle(event), theDelta as MutableCompound)
+                else expr.getArray(event).forEach { set(event, it, theDelta as MutableCompound) }
+            }
+            Changer.ChangeMode.ADD -> {
+                if (theDelta == null) rhsError()
+                @Suppress("UNCHECKED_CAST")
+                if (expr.isSingle) add(event, expr.getSingle(event), theDelta as MutableCompound)
+                else expr.getArray(event).forEach { add(event, it, theDelta as MutableCompound) }
+            }
+            Changer.ChangeMode.REMOVE -> {
+                if (theDelta == null) rhsError()
+                if (expr.isSingle) remove(event, expr.getSingle(event), theDelta)
+                else expr.getArray(event).forEach { remove(event, it, theDelta) }
+            }
+            else -> super.change(event, delta, mode)
+        }
+    }
+
+    private fun set(event: Event?, obj: Any?, delta: MutableCompound) = with (SkriptNbt.platform) {
+        when (obj) {
+            is Entity -> obj.setNbt(nbtTagOf(delta) as NbtCompound)
+            is String -> {
+                if (isFromFile) nbtToFile(Path(obj), nbtTagOf(delta) as NbtCompound)
+            }
+            else -> noNbtError(event)
```

Comment:
Unfortunately `set nbt of {_x} to {_y}` cannot know the type of {_y} until runtime, so the best way to handle it is to silently fail or print a user-friendly message. Exceptions and stacktraces will simply convince the user that there is a bug in the addon.